Sources/DeepLearning/Operators.swift

```diff
@@ -26,3 +26,326 @@ public func round<Scalar: BinaryFloatingPoint>(_ x: Tensor<Scalar>) -> Tensor<Sc
+        case .valid: return .valid
+        }
+    }
+}
+
+extension Tensor where Scalar : TensorFlowFloatingPoint {
+    /// TensorFlow builtin conv2d gradient helper for the input.
+    @inlinable
+    @differentiable(
+        wrt: (filter, backpropOutput),
+        vjp: _vjpTFConv2DBackpropInput(_:_:_:_:_:)
+    )
+    func _TFConv2DBackpropInput(
+        shape: Tensor<Int32>,
+        filter: Tensor,
+        backpropOutput: Tensor,
+        strides: (Int32, Int32, Int32, Int32),
+        padding: Padding
+    ) -> Tensor {
+        return Raw.conv2DBackpropInput(
+            inputSizes: shape,
+            filter: filter,
+            outBackprop: backpropOutput,
+            strides: [strides.0, strides.1, strides.2, strides.3],
+            padding: padding.raw)
+    }
+
+    /// TensorFlow builtin conv2d gradient helper for the filter.
+    @inlinable
+    @differentiable(
+        wrt: (input, backpropOutput),
+        vjp: _vjpTFConv2DBackpropFilter(_:_:_:_:_:)
+    )
+    func _TFConv2DBackpropFilter(
+        input: Tensor,
+        filterSizes: Tensor<Int32>,
+        backpropOutput: Tensor,
+        strides: (Int32, Int32, Int32, Int32),
+        padding: Padding
+    ) -> Tensor {
+        return Raw.conv2DBackpropFilter(
+            input,
+            filterSizes: filterSizes,
+            outBackprop: backpropOutput,
+            strides: [strides.0, strides.1, strides.2, strides.3],
+            padding: padding.raw)
+    }
+
+    @inlinable
+    func _vjpTFConv2DBackpropInput(
+        _ shape: Tensor<Int32>,
+        _ filter: Tensor,
+        _ backpropOutput: Tensor,
+        _ strides: (Int32, Int32, Int32, Int32),
+        _ padding: Padding
+    ) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
+        let value = _TFConv2DBackpropInput(shape: shape, filter: filter,
+                                                                             backpropOutput: backpropOutput,
+                                                                             strides: strides, padding: padding)
+        return (value, { v in
+            return (
+                self._TFConv2DBackpropFilter(input: v, filterSizes: shape,
+                                                                         backpropOutput: backpropOutput,
+                                                                         strides: strides, padding: padding),
+                v.convolved2D(withFilter: filter, strides: strides, padding: padding)
+            )
+        })
+    }
+
+    @inlinable
+    func _vjpTFConv2DBackpropFilter(
+        _ input: Tensor,
+        _ filterSizes: Tensor<Int32>,
+        _ backpropOutput: Tensor,
+        _ strides: (Int32, Int32, Int32, Int32),
+        _ padding: Padding
+    ) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
+        let value = _TFConv2DBackpropFilter(input: input, filterSizes: filterSizes,
+                                                                                backpropOutput: backpropOutput,
```

Comment:
Fix indentation to align with the first argument.

Reply:
Same elsewhere.

Reply:
Fixing now - this was introduced because I had to convert from the other repo's 2-space indent standard to this repo's 4-space indent standard.

Reply:
Done

Reply:
Hmm, I'm not seeing it done. `backpropOutput:` should be aligned with `input:`, which is the first arg.

Reply:
Ah sorry I misunderstood what you meant.